EIPS/eip-3102.md

```diff
@@ -0,0 +1,189 @@
+
+### Notable changes from the hexary structure
+
+ * Account and storage tries are merged, with key length between 32 and 64 bytes;
+ * RLP is no longer used;
+ * The "leaf marker" bit used in the hex prefix is also dropped. Leaves are identified as nodes with no children;
+ * Serialized nodes are hashed, no matter how small the byte length of the serialized nodes.
+
+### The trie
+
+#### Structure
+
+The trie structure is made up of _nodes_. A node $N \equiv \left(N_{l},N_{r},N_{p},N_{v}\right)$ has the following 4 components:
+
+ * $N_{l}$ is the hash to the node's _left child_. If the node does not have a left child, then $N_{l}$ is the empty hash $H_{\varnothing}$;
+ * $N_{r}$ is the hash to the node's _right child_. If the node does not have a right child, then $N_{r}$ is the empty hash $H_{\varnothing}$;
+ * the optional $N_{p}$ is the node's _prefix_ : every key into the subtrie rooted at $N$ is prefixed by this bit string;
+ * $N_{v}$ is the _value_ stored at this node. The value is **only present in leaf nodes**.
+
+Nodes with $H_\varnothing$ as both children are called _leaf nodes_, and the remaining nodes are known as _internal nodes_.
+
+#### Accessing account's balance, nonce, code, storage root and storage slots
+
+Assuming an account $A \equiv \left(A_b, A_n, A_c, A_s\right)$ at address $A_a$, the following elements can be found at the following addresses:
+
+ * The account balance $A_b \in \mathbb{B}_{32}$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 00_2$;
+ * The account nonce $A_n \in \mathbb{B}_8$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 01_2$;
+ * The code $A_c \in \mathbb{B}$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 10_2$;
+ * The root of the storage trie $A_s$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 11_2$
+     * The storage slot number $k \in \mathbb{B}_{32}$ can be found at key $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 11_2 \space\texttt{++}\space \texttt{H}(k)$;
+
+After [EIP-2926](https://eips.ethereum.org/EIPS/eip-2926) has been rolled out, $A_c$ will represent the root of the code merkelization tree. The key accessing code chunk $c$ is $\texttt{H}(A_a)_{0..253} \space\texttt{++}\space 10_2 \space\texttt{++}\space \texttt{U256}(c)$.
+
+### Node merkelization rule
+
+Leaves and nodes that have no prefix are hashed according to the rule below:
+
+```
+internal_hash = hash(left_child_hash || right_child_hash)
+leaf_hash = hash(hash(key) || hash(leaf_value))
+```
+
+If a prefix is present, the length of the path from the root to the prefixed node is further concatenated with the output of the prefix-less rule, and hashed again:
+
+```
+internal_hash_with_prefix = hash(u256(path_length_u256 - 1) || internal_hash)
+leaf_hash_with_prefix = hash(u256(path_length_u256 - 1) || leaf_hash)
+```
+
+## Rationale
+
+### BLAKE2-256
```

Comment:
What is BLAKE2-256? Is it BLAKE-256, BLAKE2s, or even BLAKE2b?

Reply:
right, it's blake2b. I'll fix it.